### DISCUSSION TITLE
Plugin to generate a single schema file, to be consumed by Relay

### DISCUSSION THREAD

**Discussion by davidkus** [99designs/gqlgen#2255] [Show and tell]
TLDR: Plugin to generate a single GraphQL schema file for the generated server, which can then be consumed by [Relay](https://relay.dev).

In trying to get [EntGo](https://entgo.io) + gqlgen + Relay to work together, I ran into a few issues:

1. **Problem**: The EntGo gql plugin does not generate the `type Mutation`. It also only generates a strict set of `Queries` - and I had a need for some custom queries.
**Solution**: Use multiple schema files, which gqlgen happily supports. In addition to the schema file generated by entgo, I included a hand-written one that contained the mutation definitions, as well as some custom queries by extending the `Query` type, e.g.:
```
extend type Query {
  viewer: User!
}
```
2. **Problem**: Relay only takes a single schema file as input.
**Solution**: Concatenate my two schema files together, à la `cat schema1.graphql schema2.graphql > combined_schema.graphql`.
3. **Problem**: Relay has a concept of [Client Schema Extensions](https://relay.dev/docs/guides/client-schema-extensions/), and they leverage GraphQL's `extend type` to achieve this. Long story short, Relay does not send anything defined in an `extend type` to the server. So, my custom queries were not being sent. 
**Solution**: Using gqlgen's [plugin framework](https://gqlgen.com/reference/plugins/), generate a schema file based on all definitions known by the generated server. Leverages github.com/vektah/gqlparser - which is an existing dependency of gqlgen.

**The Plugin**:
```
package plugin

import (
	"os"

	"github.com/99designs/gqlgen/codegen"
	"github.com/99designs/gqlgen/plugin"
	"github.com/vektah/gqlparser/v2/formatter"
)

func New() plugin.Plugin {
	return &EffectiveSchemaPlugin{}
}

type EffectiveSchemaPlugin struct{}

func (m *EffectiveSchemaPlugin) Name() string {
	return "effectiveschemagen"
}

func (m *EffectiveSchemaPlugin) GenerateCode(data *codegen.Data) error {

	f, err := os.Create("./generated/schema.graphql")
	if err != nil {
		return err
	}
	defer f.Close()

	formatter := formatter.NewFormatter(f)
	formatter.FormatSchema(data.Schema)
	return nil
}
```

This will actually generate the "effective schema" - all `extend type`s will be merged into the base `type`.

So these two files:
```
type Query {
  users: UserConnection!
}
```
```
extend type Query {
  viewer: User!
}
```
Are merged into this single file:
```
type Query {
  users: UserConnection!
  viewer: User!
}
```

Hopefully this can save someone from a headache when using gqlgen with Relay.
